### PR TITLE
feat: add "forget password" [front & backend]

### PR DESCRIPTION
**Pull request by 1340908470**
feat: add "forget password" [front & backend]

### CODE REVIEW

**Comment by hsluoyz** [comment]
@Kininaru plz review.

File: controllers/user.go [hunk 1]
```diff
@@ -111,6 +112,68 @@ func (c *ApiController) DeleteUser() {
...
+		object.DisableVerificationCode(form.Email)
+		if len(checkResult) != 0 {
+			responseText := fmt.Sprintf("Email%s", checkResult)
+			c.ResponseError(responseText)
+			return
+		}
+	} else {
+		checkPhone := fmt.Sprintf("+%s%s", form.PhonePrefix, form.Phone)
+		checkResult := object.CheckVerificationCode(checkPhone, form.PhoneCode)
+		object.DisableVerificationCode(form.Phone)
```
Comment (Kininaru) [inline controllers/user.go:145]: We should not disable the verification code until the reset action is successfully completed.

File: web/src/auth/ForgetPage.js [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+              message: i18next.t(
+                "forgetPassword:Please input your verification code!"
+              ),
+            },
+          ]}
+        >
+          {this.state.verifyType === "email" ? (
+            <CountDownInput
+              prefix={"AuditOutlined"}
+              placeHolder={"Verify code"}
```
Comment (Kininaru) [inline web/src/auth/ForgetPage.js:195]: Use i18next here.

File: web/src/auth/ForgetPage.js [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+                this.state.application?.organizationObj.owner +
+                  "/" +
+                  this.state.application?.organizationObj.name,
+              ]}
+              coolDownTime={60}
+            />
+          ) : (
+            <CountDownInput
+              prefix={"AuditOutlined"}
+              placeHolder={"Verify code"}
```
Comment (Kininaru) [inline web/src/auth/ForgetPage.js:210]: Use i18next here.

File: web/src/auth/SelfForgetPage.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import { authConfig } from "./Auth";
+import ForgetPage from "./ForgetPage";
+
+class SelfForgetPage extends React.Component {
+  render() {
+    return (
+      <ForgetPage
+        type={"forgotPassword"}
+        applicationName={authConfig.appName}
+        account={this.props.account}
```
Comment (Kininaru) [inline web/src/auth/SelfForgetPage.js:25]: It seems like that `account` is not used in `ForgetPage`. And it is impossible to have an account here when I forgot my password.

File: web/src/auth/ForgetPage.js [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+      phoneCode: "",
+      verifyType: "phone", // "email" or "phone"
+    };
+  }
+
+  UNSAFE_componentWillMount() {
+    if (this.state.type === "forgotPassword") {
+      this.getApplication();
+    } else {
+      Util.showMessage("error", `Unknown forgot type: ${this.state.type}`);
```
Comment (Kininaru) [inline web/src/auth/ForgetPage.js:58]: i18next here

File: web/src/auth/ForgetPage.js [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+        style={{ width: "250px" }}
+        size="large"
+      >
+        <Form.Item
+          style={{ height: 0, visibility: "hidden" }}
+          name="application"
+          rules={[
+            {
+              required: true,
+              message: "Please input your application!",
```
Comment (Kininaru) [inline web/src/auth/ForgetPage.js:126]: i18next here

File: web/src/auth/ForgetPage.js [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+            },
+          ]}
+        />
+        <Form.Item
+          style={{ height: 0, visibility: "hidden" }}
+          name="organization"
+          rules={[
+            {
+              required: true,
+              message: "Please input your organization!",
```
Comment (Kininaru) [inline web/src/auth/ForgetPage.js:136]: i18next here

File: web/src/auth/ForgetPage.js [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+        >
+          <Input prefix={<UserOutlined />} placeholder={"Username"} />
+        </Form.Item>
+        <Form.Item
+          name="email" //use email instead of email/phone to adapt to RequestForm in account.go
+          validateFirst
+          hasFeedback
+          rules={[
+            {
+              message: "Please input your Email/Phone string!",
```
Comment (Kininaru) [inline web/src/auth/ForgetPage.js:158]: i18next here

File: web/src/component/CountDownInput.js [hunk 9]
```diff
@@ -72,55 +82,113 @@ export const CountDownInput = (props) => {
...
+                        visible={visible}
+                        onCancel={handleCancel}
+                        onOk={handleOk}
+                    >
+                      {renderCheck()}
+                    </Modal>
+                  </div>
+                }
+            />
+        );
```
Comment (Kininaru) [inline web/src/component/CountDownInput.js:144]: These lines repeated for 3 times, but only 1 line of them are different. I suggest we can write these lines in a more elegant way.

**Review by Kininaru — CHANGES_REQUESTED** [review]

File: web/src/auth/ForgetPage.js [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+      phoneCode: "",
+      verifyType: "phone", // "email" or "phone"
+    };
+  }
+
+  UNSAFE_componentWillMount() {
+    if (this.state.type === "forgotPassword") {
+      this.getApplication();
+    } else {
+      Util.showMessage("error", `Unknown forgot type: ${this.state.type}`);
```
Comment (Kininaru) [inline web/src/auth/ForgetPage.js:59]: Maybe we can write like this: 
```javascripts
Util.showMessage("error", i18next.t(`forget:Unknown forgot type: `) + this.state.type);
```

**Review by Kininaru — CHANGES_REQUESTED** [review]

File: web/src/auth/ForgetPage.js [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+      phoneCode: "",
+      verifyType: "phone", // "email" or "phone"
+    };
+  }
+
+  UNSAFE_componentWillMount() {
+    if (this.state.type === "forgotPassword") {
+      this.getApplication();
+    } else {
+      Util.showMessage("error", i18next.t(`forget:Unknown forgot type: `) + this.state.type);
```
Comment (Kininaru) [inline web/src/auth/ForgetPage.js:59]: Then here should be: 
```javascripts
Util.showMessage("error", i18next.t("forgetPassword:Unknown forgot type:" + this.state.type);
```

I haven't noticed that there is a key `forgetPassword` in i18n, sorry...

**Review by Kininaru — CHANGES_REQUESTED** [review]

**Review by Kininaru — APPROVED** [review]
LGTM

**Comment by hsluoyz** [comment]
@1340908470 plz provide demo link.

**Comment by hsluoyz** [comment]
@1340908470 squash all commits into one

**Comment by 1340908470** [comment]
> @1340908470 squash all commits into one

done

**Comment by Kininaru** [comment]
Backend did not check if the email or phone belong to the user.

File: controllers/user.go [hunk 12]
```diff
@@ -17,6 +17,7 @@ package controllers
 import (
 	"encoding/json"
 	"fmt"
+	"github.com/casdoor/casdoor/util"
```
Comment (hsluoyz) [inline controllers/user.go:20]: sort

File: web/src/auth/AuthBackend.js [hunk 13]
```diff
@@ -12,53 +12,69 @@
...
 }
 
 export function login(values, oAuthParams) {
-  return fetch(`${authConfig.serverUrl}/api/login${oAuthParamsToQuery(oAuthParams)}`, {
-    method: 'POST',
-    credentials: "include",
-    body: JSON.stringify(values),
-  }).then(res => res.json());
+  return fetch(
+      `${authConfig.serverUrl}/api/login${oAuthParamsToQuery(oAuthParams)}`,
```
Comment (hsluoyz) [inline web/src/auth/AuthBackend.js:58]: Don't ajust the format anywhere in this PR

File: object/init.go [hunk 14]
```diff
@@ -62,9 +62,12 @@ func initBuiltInApplication() {
 		CreatedTime:    util.GetCurrentTime(),
 		DisplayName:    "Casdoor",
 		Logo:           "https://cdn.casbin.com/logo/logo_1024x256.png",
+		HomepageUrl:    "https://casdoor.org",
```
Comment (hsluoyz) [inline object/init.go:65]: Why do you commit something that is already there?

https://github.com/casbin/casdoor/blob/0f8a46fe0bd286caa0aed7658dff22c8a67e5927/object/init.go#L53-L73

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Pull request closed without merging**